src/main.cpp

```diff
@@ -5952,6 +5952,30 @@ ThresholdState VersionBitsTipState(const Consensus::Params& params, Consensus::D
     return VersionBitsState(chainActive.Tip(), params, pos, versionbitscache);
 }
 
+int VersionBitsStateBeginningBlockHeight(const Consensus::DeploymentPos pos, const ThresholdState thresholdState)
```

Comment:
AbstracrThresholdConditionChecker has the logic to walk the chain efficiently already (including first jumping back to a multiple of the period etc), so it could use the same and call GetStateFor for every multiple-or-period block back from the tip that was passed. That would also avoid breaking the abstraction that ThresholdConditionCache gives.